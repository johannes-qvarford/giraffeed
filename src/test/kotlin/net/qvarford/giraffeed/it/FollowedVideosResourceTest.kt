package net.qvarford.giraffeed.it

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.SnapshotVerifier
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig
import au.com.origin.snapshots.serializers.ToStringSnapshotSerializer
import io.quarkus.test.junit.QuarkusTest
import io.quarkus.test.junit.mockito.InjectMock
import io.restassured.RestAssured
import net.qvarford.giraffeed.fake.InMemoryHttpClient
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.InputStream
import java.net.http.HttpClient

@QuarkusTest
class FollowedVideosResourceTest {
    @InjectMock
    lateinit var httpClient: HttpClient

    @BeforeEach
    fun setupHttpClient() {
        val resource = fun (s: String): InputStream {
            return this.javaClass.classLoader.getResourceAsStream(s)!!
        }

        val map = mapOf(
            "https://api.twitch.tv/helix/channels/followed?user_id=29943195" to resource("twitch_followed_success.json"),
            "https://api.twitch.tv/helix/videos?user_id=123456&period=day&type=archive&first=3" to resource("twitch_videos_success1.json"),
            "https://api.twitch.tv/helix/videos?user_id=789012&period=day&type=archive&first=3" to resource("twitch_videos_success2.json")
        )

        InMemoryHttpClient.create(httpClient, map)
    }

    @Test
    fun twitchVideosAreMergedFromFollowedBroadcasters() {
        val expect = Expect.of(FollowedVideosResourceTest.snapshotVerifier, FollowedVideosResourceTest::class.java.getMethod("twitchVideosAreMergedFromFollowedBroadcasters"))
        val content = RestAssured.given()
            .auth().preemptive().basic("sample", "sample")
            .`when`().get("/followed-videos/atom.xml")
            .then()
            .statusCode(200)
            .extract()
            .body()
            .asString()

        expect.serializer(ToStringSnapshotSerializer::class.java).toMatchSnapshot(content)
    }

    companion object {
        private var snapshotVerifier: SnapshotVerifier? = null

        @JvmStatic
        @BeforeAll
        fun beforeAll() {
            println(System.getProperty("java.class.path"))
            snapshotVerifier = SnapshotVerifier(PropertyResolvingSnapshotConfig(), FollowedVideosResourceTest::class.java)
        }

        @JvmStatic
        @AfterAll
        fun afterAll() {
            snapshotVerifier!!.validateSnapshots()
        }
    }
}