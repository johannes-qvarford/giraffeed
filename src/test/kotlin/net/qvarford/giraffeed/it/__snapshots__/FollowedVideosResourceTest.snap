net.qvarford.giraffeed.it.FollowedVideosResourceTest.twitchVideosAreMergedFromFollowedBroadcasters=[
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<atom:feed xmlns:atom="http://www.w3.org/2005/Atom">
    <atom:icon>https://www.twitch.tv/favicon.ico</atom:icon>
    <atom:title>Latest videos from the channels I follow</atom:title>
    <atom:entry>
        <atom:author>
            <atom:name>123456</atom:name>
        </atom:author>
        <atom:id>123456_1</atom:id>
        <atom:link href="https://example.com/video1.mp4"/>
        <atom:published>2022-02-01T10:30:00Z</atom:published>
        <atom:title>Title 1</atom:title>
        <atom:content type="text/html">&lt;a href="https://example.com/video1.mp4"&gt;
&lt;img src="https://example.com/thumbnail1.jpg" /&gt;
&lt;/a&gt;</atom:content>
    </atom:entry>
    <atom:entry>
        <atom:author>
            <atom:name>123456</atom:name>
        </atom:author>
        <atom:id>123456_2</atom:id>
        <atom:link href="https://example.com/video2.mp4"/>
        <atom:published>2022-02-02T12:45:00Z</atom:published>
        <atom:title>Title 2</atom:title>
        <atom:content type="text/html">&lt;a href="https://example.com/video2.mp4"&gt;
&lt;img src="https://example.com/thumbnail2.jpg" /&gt;
&lt;/a&gt;</atom:content>
    </atom:entry>
    <atom:entry>
        <atom:author>
            <atom:name>789012</atom:name>
        </atom:author>
        <atom:id>789012_2</atom:id>
        <atom:link href="https://example.com/video3.mp4"/>
        <atom:published>2022-02-03T16:15:00Z</atom:published>
        <atom:title>Title 3</atom:title>
        <atom:content type="text/html">&lt;a href="https://example.com/video3.mp4"&gt;
&lt;img src="https://example.com/thumbnail3.jpg" /&gt;
&lt;/a&gt;</atom:content>
    </atom:entry>
    <atom:entry>
        <atom:author>
            <atom:name>789012</atom:name>
        </atom:author>
        <atom:id>789012_2</atom:id>
        <atom:link href="https://example.com/video4.mp4"/>
        <atom:published>2022-02-04T18:30:00Z</atom:published>
        <atom:title>Title 4</atom:title>
        <atom:content type="text/html">&lt;a href="https://example.com/video4.mp4"&gt;
&lt;img src="https://example.com/thumbnail4.jpg" /&gt;
&lt;/a&gt;</atom:content>
    </atom:entry>
</atom:feed>

]